Add merchant email notifications

octaedro opened the pull request:
Fixes #5300

This PR adds a new type of admin notification. With this type of notification, it will be possible to send notes via email. The procedure to create them is the same as the other ones.

**A new type**
Now it's possible to create notes type: `email`. That kind of note is sent to a certain role (`administrator` by default). Also, it's possible to extend this functionality by adding new email templates.

**A new status**
The status `sent` was added for the notes type `email`. As soon as the note is sent it's status is set as `sent`. The other statuses work as usual, so if the CTA is activated the note status will be `actioned` (the action triggering will be [addressed in a follow-up PR](https://github.com/woocommerce/woocommerce-admin/pull/5976), here we only send the CTA link but the click is not being tracked).


### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

-   [x] I've tested using only a keyboard (no mouse)
-   [ ] I've tested using a screen reader
-   [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
-   [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots
![screenshot-one wordpress test-2020 12 30-14_26_49](https://user-images.githubusercontent.com/1314156/103371293-d6e40f00-4aad-11eb-89ef-c91fd57883f5.png)

### Detailed test instructions:
_In order to test this functionality I created the branches `add/5300_test` and `add/5300_test_plain`, with notes type `email`._

- Checkout the branch "add/5300_test" (`git checkout add/5300_test`).
- Create a zip for testing with `npm run zip:test`.
- Create a `jurassic.ninja` instance (be sure to check the `Include WP Log Viewer plugin` option in the site settings).
- Upload the plugin and activate it.
- Run the cron (this tool can help [WP Crontrol](https://wordpress.org/plugins/wp-crontrol/)).
- You should have received an email, don't open it yet.
- Go to the JN site `debug.log` file. Verify that it's empty.
- Open the email and verify that the JN site `debug.log` file says `Email opened`.

**Plain email**
- In order to force a plain email I created the branch `add/5300_test_plain` so checkout that branch.
- Create the new .zip from this branch and upload it. 
- Run the cron and verify you received the email as expected.
- No opening validation will be triggered here.

<!--- Please add a Changelog note

Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance: to readme.txt under the "unreleased" list at the top of the changelog. If none exists, please add it. Also include PR number.

If changes pertain to a package, also update CHANGELOG.md in the package's folder in a similar manner.--->

octaedro committed:
Added MerchantEmailNotifications class

octaedro committed:
Added new type and Events refactor

# Conflicts:
#	src/Events.php

octaedro committed:
Added templates

octaedro committed:
Refactored MerchantEmailNotifications and NotificationEmail

octaedro committed:
Templates refactored

octaedro committed:
Added email opening tracking endpoint

octaedro committed:
Added templates handling

octaedro committed:
Moved folder `MerchantEmailNotifications`

octaedro committed:
Fixed template extensibility

octaedro committed:
Fixed note `heading` check

octaedro committed:
Added default type in `get_template_filename`

octaedro committed:
Added tests

octaedro committed:
Added flag for functionallity

octaedro committed:
Modified URL

octaedro committed:
Added required noteTypes

octaedro committed:
Removed bypass

octaedro committed:
Fixed plain link

octaedro committed:
Fixed comment

octaedro committed:
Turned email notifications on by default

This commit adds the code to turn email notifications on by default

octaedro committed:
Fixed email styles

becdetat commented on includes/emails/html-mechant-notification.php:
```diff
@@ -0,0 +1,50 @@
+<?php
```
The file name is off, missing an `r` in `merchant`

octaedro replied:
Nice catch! This was fixed in the commit d6c6277

becdetat commented on src/Notes/MerchantEmailNotifications/MerchantEmailNotifications.php:
```diff
@@ -0,0 +1,77 @@
+	/**
+	 * Send all the notifications type `email`.
+	 */
+	public static function possible_send() {
```
This could just be called `run()`

octaedro replied:
Good idea! I renamed the method in the commit 79f811c

becdetat commented on src/Notes/MerchantEmailNotifications/MerchantEmailNotifications.php:
```diff
@@ -0,0 +1,77 @@
+			)
+		);
+
+		if ( empty( $notes ) ) {
```
This isn't really needed as the `foreach` will handle an empty array, having this makes it slightly harder to grok imo.

octaedro replied:
Good catch! Modified in the commit be9301d

becdetat commented on src/Notes/MerchantEmailNotifications/MerchantEmailNotifications.php:
```diff
@@ -0,0 +1,77 @@
+	 *
+	 * @param object $note The note to send.
+	 */
+	public static function send_mechant_notification( $note ) {
```
typo `mechant`

octaedro replied:
Thank you Bec, this was fixed in the commit 0d78e23

becdetat commented on src/Notes/MerchantEmailNotifications/MerchantEmailNotifications.php:
```diff
@@ -0,0 +1,77 @@
+	 * @param object $note The note to send.
+	 * @return array Emails to notify
+	 */
+	public static function get_emails_to_notice( $note ) {
```
This could be more descriptive, like `get_notification_email_addresses()`

octaedro replied:
I like the idea, this was modified in the commit c6663ce

becdetat commented on src/Notes/MerchantEmailNotifications/MerchantEmailNotifications.php:
```diff
@@ -0,0 +1,77 @@
+		foreach ( $users_emails as $user_email ) {
+			if ( is_email( $user_email ) ) {
+				$email = new NotificationEmail( $note );
+				$email->trigger( $user_email );
```
If the recipient in `NotificationEmail::trigger` wasn't a class property and wasn't being mutated in `trigger` then this could be simplified to:

```
$email = new NotificationEmail( $note );
foreach ( $users_emails as $user_email ) {
    if ( is_email( $user_email ) ) {
        $email->trigger( $user_email );
    }
}

becdetat replied:
This is a minor issue though, there's nothing wrong with how this is implemented.

octaedro replied:
You're right! That's a good improvement to the method. I added it in the commit b38fa44

becdetat reviewed (approved):
Everything looks good apart from a few naming niggles. It all tests well and is Awesome. Pre-approving this.

octaedro committed:
Fixed typo

octaedro committed:
Renamed method "possible_send" as "run"

octaedro committed:
Removed unnecessary control

octaedro committed:
Fixed another typo

octaedro committed:
Renamed method as "get_notification_email_addresses"

octaedro committed:
Refactored method "send_merchant_notification"

octaedro committed:
Renamed plain-merchant-notification

octaedro committed:
Fixed tests

octaedro commented:
Thank you @becdetat  for your review. I rebased the PR and applied the suggested modifications.

octaedro committed:
Merchant email notifications - Action triggering (#5976)

* Added templates

# Conflicts:
#	includes/emails/plain-mechant-notification.php

# Conflicts:
#	includes/emails/html-merchant-notification.php

* Added note action triggering

This commit adds the note actions triggering

# Conflicts:
#	includes/emails/html-merchant-notification.php

* Fixed error handling

* Refactored "trigger_note_action" method

* Fixed actions url

* Fixed note URL

* Added external redirect

* Added image and html handling for email

* Fixed tests

* Fixed buttons style

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>

octaedro committed:
Add your first product: email notification (#6024)

* Added AddFirstProduct note

# Conflicts:
#	src/Events.php

* Added "AddFirstProduct" email note

This commit adds the email note "AddFirstProduct"

* Fixed image

This commit removes the image img-product-light.svg to use a png instead. Otherwise, the gmail proxy would break the image

* Fixed readme.txt

Co-authored-by: Fernando Marichal <contacto@fernandomarichal.com>

octaedro committed:
Added readme.txt

octaedro committed:
Merge branch 'add/5300' of github.com:woocommerce/woocommerce-admin into add/5300

# Conflicts:
#	readme.txt

octaedro committed:
Merge branch 'main' into add/5300
